### PR TITLE
Migrate DetermineJoinDistributionType to iterative optimizer

### PR DESCRIPTION
**Pull request by findepi**
This migrates `DetermineJoinDistributionType` legacy optimizer into iterative optimizer's rules.
Since semi-join requires special handling when in a `DELETE` query, there is a new legacy optimizer left that takes care just of this case. This one is due to be migrated when we have required infrastructure (traits).

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    private static boolean mustBroadcastJoin(JoinNode node, Lookup lookup)
+    {
+        return isAtMostScalar(node.getRight(), lookup) || isCrossJoin(node);
+    }
+
+    private static boolean isCrossJoin(JoinNode node)
+    {
+        return node.getType() == INNER && node.getCriteria().isEmpty();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: use `JoinNode#isCrossJoin` instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java [hunk 2]
```diff
@@ -27,13 +28,10 @@
...
 
-import static com.facebook.presto.SystemSessionProperties.isDistributedJoinEnabled;
-import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isAtMostScalar;
-import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
-import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
-import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+import static com.facebook.presto.sql.planner.iterative.Lookup.noLookup;
 import static java.util.Objects.requireNonNull;
 
+@Deprecated
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java]: This hasn't the same functionality as `iterative/rule/DetermineJoinDistributionType.java` (it handles delete node). Consider either refactoring this class to `ReplicateJoinUnderDelete` to be consistent with semi join. I don't see much value in keeping this class.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ReplicateSemiJoinUnderDelete.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.Map;
+
+import static com.facebook.presto.sql.planner.plan.SemiJoinNode.DistributionType.REPLICATED;
+
+/*
+ * For delete queries, the TableScan node that corresponds to the table being deleted
+ * must be collocated with the Delete node, so any semi-join in between must be replicated
+ */
+// TODO migrate to traits
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ReplicateSemiJoinUnderDelete.java]: Ideally I think `DeleteNode` should have some kind of `SOURCE` data distribution requirement. This requirement would enforce that data is never transmitted over network.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SemiJoinNode.java [hunk 4]
```diff
@@ -140,7 +140,7 @@ public Symbol getSemiJoinOutput()
     }
 
     @Override
-    public PlanNode replaceChildren(List<PlanNode> newChildren)
+    public SemiJoinNode replaceChildren(List<PlanNode> newChildren)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SemiJoinNode.java]: Unrelated change. It's not consistent with `JoinNode::replaceChildren`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: You might want to update `com.facebook.presto.sql.planner.optimizations.DetermineJoinDistributionType.Rewriter#isCrossJoin`

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: `JoinNode#isCrossJoin` returns `true` when join doesn't have equi criteria but just a filter. It would instruct the rule to pick the partitioned join which makes no sense.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SemiJoinNode.java (hunk 4 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SemiJoinNode.java]: It's related, used in `ReplicateSemiJoinUnderDelete`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        }
+
+        if (isDistributedJoinEnabled(context.getSession())) {
+            return PARTITIONED;
+        }
+
+        return REPLICATED;
+    }
+
+    private static boolean mustBroadcastJoin(JoinNode node, Context context)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: rename to `preferredBroadcastJoin`?

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: this is no longer compatible with `com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java`

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: Correct.

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: Inlined to avoid confusing names

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java (hunk 2 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DetermineJoinDistributionType.java]: It is equivalent but keeping it around allows disabling the new rule.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java [hunk 6]
```diff
@@ -233,6 +233,11 @@ public PlanNode replaceChildren(List<PlanNode> newChildren)
         return new JoinNode(getId(), type, newLeft, newRight, criteria, newOutputSymbols, filter, leftHashSymbol, rightHashSymbol, distributionType);
     }
 
+    public JoinNode withDistributionType(DistributionType distributionType)
+    {
+        return new JoinNode(getId(), type, left, right, criteria, outputSymbols, filter, leftHashSymbol, rightHashSymbol, Optional.of(distributionType));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: put each arg in separate line

**Comment by findepi** [comment]
Superseded by https://github.com/prestodb/presto/pull/9747

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+            // The implementation of full outer join only works if the data is hash partitioned.
+            return PARTITIONED;
+        }
+
+        if (node.getCriteria().isEmpty() && type == INNER) {
+            // There is nothing to partition on
+            return REPLICATED;
+        }
+
+        if (isAtMostScalar(node.getRight(), context.getLookup())) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: I think conditions which returns `REPLICATED` should be tested before `PARTITIONED` ones

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+import static com.facebook.presto.sql.planner.iterative.rule.test.PlanBuilder.expressions;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.PARTITIONED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.FULL;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.LEFT;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.RIGHT;
+
+@Test(singleThreaded = true)
+public class TestDetermineJoinDistributionType
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: can you add a test that when `EnforceSingleRow` is used on the right, then join is REPLICATED always?

Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: This would be testing `isAtMostScalar` right?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 7 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: You mean changing logic or just order?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java (hunk 6 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java]: still under line length limit (just as 6 lines above)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java (hunk 8 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: yes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 7 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: changing the logic

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java]: You can't to `REPLICATED` (broadcast) for RIGHT/FULL join, even if right side is at most scalar.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java (hunk 8 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestDetermineJoinDistributionType.java]: at your service
